Set build_jobs dynamically in CI to avoid oversubscription

haampie opened the pull request:
Instead of setting a fixed number of jobs (32) in spack,
set them dynamically based on YAV (yet another variable),
that corresponds to the kubernetes request.

This is mostly for packages that use a lot of parallel
resources for a long time, like paraview and what not,
since as it is right now, their resource request is 11 CPUs
and their usages is 3x as much, so, it's likely that jobs
get killed because of oversubscription.

Notice, I've increased the resource request from 11 to
16 in "huge", because 11 is not huge.

For small packages this doesn't matter much, since
often the peak resource usage is very short. But I've
also added a build_jobs limit there anyhow... if that
slows down builds too much, we can allow more
oversubscription there again later.

scottwittenburg reviewed (dismissed):
This seems like a good idea, but as you point out, while it hopefully addresses the problems we've seen with `py-tensorflow`, it may have some detrimental effect on other specs.  But it's worth giving a try, IMO.

scottwittenburg commented:
Just wondering if below is something we might consider in the future, to have more control over the `-j` factor for groups of specs.  I'm not sure if the `spack config ...` incantation is correct, but hopefully the idea is clear enough.

```
diff --git a/share/spack/gitlab/cloud_pipelines/stacks/e4s/spack.yaml b/share/spack/gitlab/cloud_pipelines/stacks/e4s/spack.yaml
index 8b961a32e0..2a632ac5b2 100644
--- a/share/spack/gitlab/cloud_pipelines/stacks/e4s/spack.yaml
+++ b/share/spack/gitlab/cloud_pipelines/stacks/e4s/spack.yaml
@@ -252,6 +252,7 @@ spack:
       - spack arch
       - cd ${SPACK_CONCRETE_ENV_DIR}
       - spack env activate --without-view .
+      - spack config add "config:build_jobs:${SPACK_CONFIG_BUILD_JOBS}"
       - spack config add "config:install_tree:projections:${SPACK_JOB_SPEC_PKG_NAME}:'morepadding/{architecture}/{compiler.name}-{compiler.version}/{name}-{version}-{hash}'"
       - mkdir -p ${SPACK_ARTIFACTS_ROOT}/user_data
       # AWS runners mount E4S public key (verification), UO runners mount public/private (signing/verification)
@@ -284,6 +285,7 @@ spack:
             CI_JOB_SIZE: huge
             KUBERNETES_CPU_REQUEST: 11000m
             KUBERNETES_MEMORY_REQUEST: 42G
+            SPACK_CONFIG_BUILD_JOBS: 12
 
       - match:
           - cuda
@@ -321,6 +323,7 @@ spack:
             CI_JOB_SIZE: large
             KUBERNETES_CPU_REQUEST: 8000m
             KUBERNETES_MEMORY_REQUEST: 12G
+            SPACK_CONFIG_BUILD_JOBS: 8
 
       - match:
           - adios2
@@ -389,6 +392,7 @@ spack:
             CI_JOB_SIZE: "medium"
             KUBERNETES_CPU_REQUEST: "2000m"
             KUBERNETES_MEMORY_REQUEST: "4G"
+            SPACK_CONFIG_BUILD_JOBS: 2
 
       - match:
           - alsa-lib
@@ -450,12 +454,14 @@ spack:
             CI_JOB_SIZE: "small"
             KUBERNETES_CPU_REQUEST: "500m"
             KUBERNETES_MEMORY_REQUEST: "500M"
+            SPACK_CONFIG_BUILD_JOBS: 1
 
       - match: ['os=ubuntu20.04']
         runner-attributes:
           tags: ["spack", "x86_64"]
           variables:
             CI_JOB_SIZE: "default"
+            SPACK_CONFIG_BUILD_JOBS: 1
 
     broken-specs-url: "s3://spack-binaries/broken-specs"
```

haampie commented:
If there's an alternative way of communicating the number of usable cpus to spack, that'd be helpful.

haampie commented:
Yeah, the main downside is capping the number of jobs is probably pessimistic, since it's likely the majority of the builds of many packages is on average not very parallel (including git clone, configure, cmake, etc).

kwryankrattiger commented:
#34272 is just about to be merged, I think it will be better to wait for that to put this in.

haampie commented:
Okay, I dropped the fixed `build_jobs` from the environments, made it "dynamic' with `spack config add`, but also added more jobs to those "huge" jobs cause they weren't particularly huge -- it looks like all the tensorflow and paraview builds etc kinda saturate the number of jobs during their build, so 11 is not "huge".

scottwittenburg reviewed (dismissed):
:tada:

haampie committed:
set config:build_jobs dynamically

kwryankrattiger reviewed (dismissed):
LGTM, hopefully this gets things to pass now!

haampie committed:
trigger a rebuild

haampie committed:
Merge branch 'develop' into fix/oversubscription-in-ci

haampie commented:
@kwryankrattiger / @scottwittenburg / @zackgalbreath can you take this PR over from here?

You still gotta push `git revert a0d68a38a899286ef9a589de25c4f94159f9cae2` of course ;)

zackgalbreath committed:
Merge branch 'develop' into fix/oversubscription-in-ci

haampie commented:
> the resource "CPURequest" requested "16000m" is higher than limit allowed "12"

kwryankrattiger committed:
CI: Drop huge CPU request to 12

kwryankrattiger commented:
I drop the CPU request to 12. If the job times out then maybe getting kube to give more than 12 cores would be another fix.

haampie commented:
Seems like it works. Wanna revert that commit and merge?

haampie committed:
Revert "trigger a rebuild"

This reverts commit a0d68a38a899286ef9a589de25c4f94159f9cae2.
